### PR TITLE
trace sessions and records command

### PR DESCRIPTION
**Pull request by nicolasparada**
# Summary of this proposal

Adding new commands:

 - `calyptia create trace_session --pipeline PIPELINE`
 - `calyptia get trace_sessions --pipeline PIPELINE`
 - `calyptia get trace_session SESSION`
 - `calyptia get trace_session --pipeline PIPELINE`
 - `calyptia delete trace_session --pipeline PIPELINE`
 - `calyptia get trace_records --session SESSION_ID`

Also did a couple of changes on how we do confirmation of deletes. Now is `N` (no) by default.

And removed the constrain to validate the core image tag from the index.
The index is just used for autocompletion now, but we allow any tag.

## Notes for the reviewer

Still missing some tests, but the commands to work with sessions were manually tested and they work.
The command to list the trace records hasn't been tested yet.

Ideally you should be able to start a tracing session. And then see the records.

```bash
calyptia create trace_session --pipeline PIPELINE
calyptia get trace_records --session SESSION_ID
```

For that we need an instance using the latest core image with tracing support.
But let's merge as it is and fix remaining stuff once we get it working.

## Checklist for submitter

- [ ] My PR has a related issue/bug number.
- [ ] My PR provides tests
  - [ ] Integration tests are added/passes
  - [ ] Unit tests are added/passes
  - [ ] End-to-end tests added/passes
  - [ ] Static code check added/passes
  - [ ] Linting on documentation added/passes
  - [ ] Does not affect code coverage stats
- [ ] My PR requires updating dependencies
- [ ] My PR has the documentation changes required.

## Checklist for reviewer

- [ ] The proposal fixes a bug/issue or implements a new feature that is well described.
- [ ] The proposal has sufficient test cases that covers the changes.
  - [ ] If changes an API, it does not break backwards compatibility (-1 major version).
- [ ] If integration is required, the proposal has integration tests.
- [ ] The proposal does not break coverage stats
- [ ] The proposal has the required documentation changes.

## Backport

- [ ] Backport to the latest stable release.

### CODE REVIEW

File: cmd/calyptia/create_core_instance_k8s.go [hunk 1]
```diff
@@ -36,12 +37,7 @@ func newCmdCreateCoreInstanceOnK8s(config *config, testClientSet kubernetes.Inte
 		RunE: func(cmd *cobra.Command, args []string) error {
 			ctx := context.Background()
 
-			coreDockerImageTag, err := FindCoreImageTag(ctx, coreInstanceVersion)
-			if err != nil {
-				return fmt.Errorf("could not find a matching container image for version: %v, %w", coreInstanceVersion, err)
-			}
-
-			coreDockerImage := fmt.Sprintf("%s:%s", defaultCoreDockerImage, coreDockerImageTag)
+			coreDockerImage := fmt.Sprintf("%s:%s", defaultCoreDockerImage, coreInstanceVersion)
```
Comment (nicolasparada) [inline cmd/calyptia/create_core_instance_k8s.go:40]: I removed the image tag constraint.

File: cmd/calyptia/delete_core_instance_aws.go [hunk 2]
```diff
@@ -88,8 +91,16 @@ func newCmdDeleteCoreInstanceOnAWS(config *config, client awsclient.Client) *cob
 			fmt.Fprintln(cmd.OutOrStdout(), "The following resources will be removed from your AWS account:\n"+strings.Join(toDelete, "\n"))
 
 			if !confirmDelete && !isNonInteractiveMode {
-				cmd.Println("\nYou confirm the deletion of those resources? [Y/n]")
-				confirmDelete = ask(cmd.InOrStdin(), cmd.ErrOrStderr())
+				cmd.Print("You confirm the deletion of those resources? [y/N] ")
```
Comment (nicolasparada) [inline cmd/calyptia/delete_core_instance_aws.go:94]: Changed how we confirm. `N` (no) by default.

File: cmd/calyptia/delete_environment.go [hunk 3]
```diff
@@ -27,14 +28,19 @@ func newCmdDeleteEnvironment(c *config) *cobra.Command {
...
 			}
 			environment := environments.Items[0]
-			if !confirmDelete && !isNonInteractiveMode {
-				cmd.Println("This will remove ALL your agents, aggregators. Do you confirm? [Y/n]")
-				confirmDelete = ask(cmd.InOrStdin(), cmd.ErrOrStderr())
-			}
 			if !confirmDelete {
-				cmd.Println("operation canceled")
-				return nil
+				cmd.Print("This will remove ALL your agents, aggregators. Do you confirm? [y/N] ")
```
Comment (nicolasparada) [inline cmd/calyptia/delete_environment.go:32]: Changed how we read confirm. No by default.

File: cmd/calyptia/update_core_instance_k8s.go [hunk 4]
```diff
@@ -26,12 +27,7 @@ func newCmdUpdateCoreInstanceK8s(config *config, testClientSet kubernetes.Interf
 			ctx := context.Background()
 			aggregatorKey := args[0]
 
-			coreDockerImageTag, err := FindCoreImageTag(ctx, newVersion)
-			if err != nil {
-				return fmt.Errorf("could not find a matching container image for version: %v, %w", newVersion, err)
-			}
-
-			coreDockerImage := fmt.Sprintf("%s:%s", defaultCoreDockerImage, coreDockerImageTag)
+			coreDockerImage := fmt.Sprintf("%s:%s", defaultCoreDockerImage, newVersion)
```
Comment (nicolasparada) [inline cmd/calyptia/update_core_instance_k8s.go:30]: Remove image tag constraint.

File: cmd/calyptia/get_trace_session.go [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+	g, gctx := errgroup.WithContext(config.ctx)
+	for _, pip := range pp {
+		a := pip
+		g.Go(func() error {
+			got, err := config.cloud.TraceSessions(gctx, a.ID, types.TraceSessionsParams{})
+			if err != nil {
+				return err
+			}
+
+			mu.Lock()
```
Comment (pwhelan) [inline cmd/calyptia/get_trace_session.go:237]: Do we need to lock access to the local variables?

**Review by pwhelan — CHANGES_REQUESTED** [review]
All I saw with this check was the use of a mutex, which I find excessive for single threaded client programs.

File: cmd/calyptia/get_trace_session.go (hunk 5 above)
Comment (nicolasparada) [reply cmd/calyptia/get_trace_session.go:237]: Yeah, because the data is fetched concurrently.
We do the same for all of these completions.

**Review by niedbalski — APPROVED** [review]

**Pull request merged**